src/core/execute.c

```diff
@@ -720,74 +720,129 @@ static int ask_for_confirmation(char *response, char **argv) {
+                const char *g;
+
+                if (k >= ngroups_max) {
+                        free(l_gids);
+                        return -E2BIG;
+                }
+
+                g = *i;
+                r = get_group_creds(&g, l_gids+k);
+                if (r < 0) {
+                        free(l_gids);
+                        return r;
+                }
+
+                k++;
         }
 
-        if (context->supplementary_groups) {
-                int ngroups_max, k;
-                gid_t *gids;
-                char **i;
+        if (k >= ngroups_max) {
+                free(l_gids);
+                return -E2BIG;
+        }
 
-                /* Final step, initialize any manually set supplementary groups */
-                assert_se((ngroups_max = (int) sysconf(_SC_NGROUPS_MAX)) > 0);
+        /* Close gids and GID_INVALID can't be used from user passed values
+         * in any case we will fail before if we can't resolve a group */
+        l_gids[k] = GID_INVALID;
 
-                if (!(gids = new(gid_t, ngroups_max)))
-                        return -ENOMEM;
+        *supplementary_gids = l_gids;
 
-                if (keep_groups) {
-                        k = getgroups(ngroups_max, gids);
-                        if (k < 0) {
-                                free(gids);
-                                return -errno;
-                        }
-                } else
-                        k = 0;
+        return 0;
+}
 
-                STRV_FOREACH(i, context->supplementary_groups) {
-                        const char *g;
+static int enforce_groups(const ExecContext *context, const char *user,
+                          const char *group, gid_t gid, gid_t *supplementary_gids) {
+        int r;
 
-                        if (k >= ngroups_max) {
-                                free(gids);
-                                return -E2BIG;
-                        }
+        assert(context);
 
-                        g = *i;
-                        r = get_group_creds(&g, gids+k);
-                        if (r < 0) {
-                                free(gids);
-                                return r;
-                        }
+        /* Set GID and supplementary group list. */
+        if (supplementary_gids) {
+                unsigned i = 0;
 
-                        k++;
-                }
+                while (supplementary_gids[i] != GID_INVALID)
+                        i++;
 
-                r = maybe_setgroups(k, gids);
-                if (r < 0) {
-                        free(gids);
+                /* Final step, manually set supplementary groups */
+                r = maybe_setgroups(i, supplementary_gids);
+                if (r < 0)
                         return r;
-                }
 
-                free(gids);
+                /* Lets free this here also */
+                supplementary_gids = mfree(supplementary_gids);
+        }
+
+        if (group || user) {
```

Comment:
wouldn't it be nicer to check `if (gid_is_valid(gid))` here instead? Should hold in the same cases, but would be nicer since `gid` is the only var we actually care about here...